extension/services/ExtensionMetadataReader.cfc

```diff
@@ -0,0 +1,327 @@
+			if ( arguments.version == "latest" || isEmpty( arguments.version ) ) {
+				arguments.version = mapped[ arguments.id ]._latest;
+			}
+			if ( StructKeyExists( mapped[ arguments.id ], arguments.version ) ) {
+				var ext = StructCopy( mapped[ arguments.id ][ arguments.version ] );
+				if ( !arguments.withLogo ) {
+					ext.image = "";
+				}
+				return ext;
+			}
+		}
+
+		return {};
+	}
+
+// PRIVATE HELPERS
+	private function _listLexFilesFromBucket() {
+		return DirectoryList(
+			  path     = getS3Root()
+			, sort     = "name"
+			, listInfo = "query"
+			, filter   = "*.lex"
+		);
+	}
+
+	private function _readExistingMetaFileFromS3() {
+		var metaFile = getS3Root() & "/extensions.json";
+		if ( FileExists( metaFile  ) ) {
+			return DeserializeJson( FileRead( metaFile ), false );
+		}
+
+		return _getEmptyExtensionsQuery();
+	}
+
+	private function _writeMetaFileToS3( meta ) {
+		var metaFile = getS3Root() & "/extensions.json";
+
+		FileWrite( metaFile, SerializeJson( meta ) );
+	}
+
+	private function _getEmptyExtensionsQuery() {
+		return QueryNew( "id,version,versionSortable,name,description,filename,image,category,author,created,releaseType,minLoaderVersion,minCoreVersion,price,currency,disableFull,trial,older,olderName,olderDate,promotionLevel,promotionText,projectUrl,sourceUrl,documentionUrl" );
+	}
+
+	private function _mapExtensionQueryByFilename( extensionQuery ) {
+		var mapped = {};
+
+		for( var e in arguments.extensionQuery ) {
+			mapped[ e.filename ] = e;
+		}
+
+		return mapped;
+	}
+
+	private function _addLexFile( fileName, extensionMetaQuery ) {
+		SystemOutput( "Loading new extension file [#arguments.fileName#] from S3.", true );
+
+		try {
+			var tmpFile  = _copyRemoteFileToTmpFile( arguments.fileName );
+			var qryCols  = ListToArray( arguments.extensionMetaQuery.columnList );
+			var extMeta  = _initExtensionMetaFromManifest( tmpFile, qryCols );
+
+			extMeta.filename        = arguments.fileName;
+			extMeta.versionSortable = Len( extMeta.version ) ? Util::sortableVersionString( extMeta.version ) : "";
+			extMeta.trial           = false; // "TODO"???
+			extMeta.releaseType     = Len( extMeta.releaseType ) ? extMeta.releaseType : "all";
+			extMeta.image           = _getLogoThumbnail( tmpFile );
+
+			QueryAddRow( arguments.extensionMetaQuery, extMeta );
+
+			_processExtensionJars( tmpFile );
+
+			try {
+				FileDelete( tmpFile );
+			} catch( any e ) {
+				SystemOutput( e, true );
+			}
+		} catch( any e ) {
+			// for now, just to get through them all!
+			SystemOutput( "Error processing lex file: [#arguments.fileName#]", true );
+		}
+	}
+
+	private function _readLexManifest( filePath ) {
+		try {
+			var mf = ManifestRead( arguments.filePath );
+		} catch( any e ) {}
+
+		return mf.main ?: {};
+	}
+
+	private function _copyRemoteFileToTmpFile( fileName ) {
+		var remoteFile = getS3Root() & "/" & arguments.fileName;
+		var tmpFile = GetTempDirectory() & "/" & arguments.fileName;
```

Comment:
let's use https://docs.lucee.org/reference/functions/gettempfile.html here, to avoid any conflicts

`GetTempFile( dir=string, prefix=string, extension=string );`